### PR TITLE
Issue 38 W.I.P

### PR DESCRIPTION
**Pull request by jcalderonzumba**
So for the current work on test fixing and removing the self made replacement tests

### CODE REVIEW

**Comment by jcalderonzumba** [comment]
@stof Ok done.

**Comment by jcalderonzumba** [comment]
So now situation is 8 errors (Javascript related) and 5 test failures:
Out of this 5 test failures 3 are related purely to the way phantomjs returns the HTML of a page, let me give an example:

```
2) Behat\Mink\Tests\Driver\Basic\HeaderTest::testHeaders
Failed asserting that 'Array
(
    [DOCUMENT_ROOT] =&gt; /Users/juan/code/scm/pjsdriver/vendor/behat/mink/driver-testsuite/web-fixtures
    [REMOTE_ADDR] =&gt; 127.0.0.1
    [REMOTE_PORT] =&gt; 59645
    [SERVER_SOFTWARE] =&gt; PHP 5.5.20 Development Server
    [SERVER_PROTOCOL] =&gt; HTTP/1.1
    [SERVER_NAME] =&gt; 127.0.0.1
    [SERVER_PORT] =&gt; 6789
    [REQUEST_URI] =&gt; /headers.php
    [REQUEST_METHOD] =&gt; GET
    [SCRIPT_NAME] =&gt; /headers.php
    [SCRIPT_FILENAME] =&gt; /Users/juan/code/scm/pjsdriver/vendor/behat/mink/driver-testsuite/web-fixtures/headers.php
    [PHP_SELF] =&gt; /headers.php
    [HTTP_USER_AGENT] =&gt; Mozilla/5.0 (Macintosh; Intel Mac OS X) AppleWebKit/534.34 (KHTML, like Gecko) PhantomJS/1.9.7 Safari/534.34
    [HTTP_ACCEPT] =&gt; text/html,application/xhtml+xml,application/xml;q=0.9,*/*;q=0.8
    [HTTP_ACCEPT_LANGUAGE] =&gt; fr
    [HTTP_CONNECTION] =&gt; Keep-Alive
    [HTTP_ACCEPT_ENCODING] =&gt; gzip
    [HTTP_HOST] =&gt; 127.0.0.1:6789
    [REQUEST_TIME_FLOAT] =&gt; 1449764143.4814
    [REQUEST_TIME] =&gt; 1449764143
)' contains "[HTTP_ACCEPT_LANGUAGE] => fr".
```
HTTP_ACCEPT_LANGUAGE = fr does exists, the problem is that phantomjs as a real browser "auto-fixes" the data it receives from the servers and returns content enclosed in an HTML/XML element.

See: [http://phantomjs.org/api/webpage/property/content.html](http://phantomjs.org/api/webpage/property/content.html)

In your test suite the file ``headers.php`` has the following content:

```
<!DOCTYPE html>
<html>
<head>
    <title>Headers page</title>
    <meta http-equiv="Content-Type" content="text/html;charset=UTF-8"/>
</head>
<body>
    <?php print_r($_SERVER); ?>
</body>
</html>
```

Then internally phantomJS see this as:
```
<!DOCTYPE html><html><head>
    <title>Headers page</title>
    <meta http-equiv="Content-Type" content="text/html;charset=UTF-8">
</head>
<body>
    Array
(
    [DOCUMENT_ROOT] =&gt; /Users/juan/code/scm/pjsdriver/vendor/behat/mink/driver-testsuite/web-fixtures
    [REMOTE_ADDR] =&gt; 127.0.0.1
    [REMOTE_PORT] =&gt; 61629
    [SERVER_SOFTWARE] =&gt; PHP 5.5.20 Development Server
    [SERVER_PROTOCOL] =&gt; HTTP/1.1
    [SERVER_NAME] =&gt; 127.0.0.1
    [SERVER_PORT] =&gt; 6789
    [REQUEST_URI] =&gt; /headers.php
    [REQUEST_METHOD] =&gt; GET
    [SCRIPT_NAME] =&gt; /headers.php
    [SCRIPT_FILENAME] =&gt; /Users/juan/code/scm/pjsdriver/vendor/behat/mink/driver-testsuite/web-fixtures/headers.php
    [PHP_SELF] =&gt; /headers.php
    [HTTP_USER_AGENT] =&gt; Mozilla/5.0 (Macintosh; Intel Mac OS X) AppleWebKit/534.34 (KHTML, like Gecko) PhantomJS/1.9.7 Safari/534.34
    [HTTP_ACCEPT] =&gt; text/html,application/xhtml+xml,application/xml;q=0.9,*/*;q=0.8
    [HTTP_X_MINK_TEST] =&gt; test
    [HTTP_CONNECTION] =&gt; Keep-Alive
    [HTTP_ACCEPT_ENCODING] =&gt; gzip
    [HTTP_ACCEPT_LANGUAGE] =&gt; en-US,*
    [HTTP_HOST] =&gt; 127.0.0.1:6789
    [REQUEST_TIME_FLOAT] =&gt; 1449764706.448
    [REQUEST_TIME] =&gt; 1449764706
)


</body></html>
```

Looks like is decoding entities, as of today there's NO way to get the RAW content without phantomjs fixing data by itself.

In my opinion PhantomJS is acting correctly as =&gt is => and > is a HTML tag therefore it should be encoded. Actually this behaviour is not problem of PhantomJS but of QT library. 

I'm a C and C++ guy so i dig deep and the "magic" is done here:
[https://github.com/qtproject/qtwebkit/blob/dev/Source/WebCore/editing/MarkupAccumulator.cpp](https://github.com/qtproject/qtwebkit/blob/dev/Source/WebCore/editing/MarkupAccumulator.cpp]

@stof What do you think about this because doing entities decoding on our side does not seem like a good idea.

**Comment by stof** [comment]
@jcalderonzumba the issue is related to the fact that some of our responses are actually plaintext ones, not html ones, but we may not send the right content type.

**Comment by jcalderonzumba** [comment]
Driver is close to actually be good on the test suite, javascript has been fixed finally. Errors still on the html entities encoding and some things related to basic Auth.

@stof: I don't see a quick easy solution for the html entities problems other than modifying the test suite itself.

**Comment by jcalderonzumba** [comment]
@stof Sorry to bother now i have another doubt regarding driver and tests fixes:

This function is part of the driver: 
```php
public function resizeWindow($width, $height, $name = null);
```

Question is: Is this function objective to measure real browser window controlled by Selenium for example because phantomJS is headless meaning that the following code:
```php
  public function testResizeWindow()
    {
        $this->getSession()->visit($this->pathTo('/index.html'));
        $session = $this->getSession();

        $session->resizeWindow(400, 300);
        $session->wait(1000, 'false');

        $script = 'return Math.abs(window.outerHeight - 300) <= 100 && Math.abs(window.outerWidth - 400) <= 100;';

        $this->assertTrue($session->evaluateScript($script));
    }
```

Will fail in this driver because ``window.outer*`` does not exist and therefore is 0. On the other hand phantomjs does allow to "resize" your browser by: [http://phantomjs.org/api/webpage/property/viewport-size.html](http://phantomjs.org/api/webpage/property/viewport-size.html)

This means that ``window.innerHeight and window.innerWidth`` will have the new size after "resizing" via viewport.

For more information check: 
[http://stackoverflow.com/questions/3437786/get-the-size-of-the-screen-current-web-page-and-browser-window](http://stackoverflow.com/questions/3437786/get-the-size-of-the-screen-current-web-page-and-browser-window)
[https://developer.mozilla.org/en-US/docs/Web/API/Window/outerHeight](https://developer.mozilla.org/en-US/docs/Web/API/Window/outerHeight)

Shall i make a PR to Mink to take this into consideration else this particular test will never pass?

**Comment by stof** [comment]
@jcalderonzumba the test may need to be improved, but you would have to try it against Selenium first.

**Comment by jcalderonzumba** [comment]
OK, i will take a look, this could be a cascade check.

First check the outer* as it is now if that goes true then no more check, else check for the following approach.

```js
var w = window,
    d = document,
    e = d.documentElement,
    g = d.getElementsByTagName('body')[0],
    x = w.innerWidth || e.clientWidth || g.clientWidth,
    y = w.innerHeight|| e.clientHeight|| g.clientHeight;
```

I'll make a PR to that particular test, thanks.

File: src/JavascriptTrait.php [hunk 1]
```diff
@@ -11,11 +11,30 @@
 trait JavascriptTrait {
 
   /**
+   * Helper function to fix javascript code before sending it to the phantomjs API
+   * @param $script
+   * @return string
+   */
+  protected function fixJavascriptForUse($script){
```
Comment (stof) [inline src/JavascriptTrait.php]: this should probably be private

**Comment by jcalderonzumba** [comment]
closing this as the new driver test suite is integrated and now driver is almost ready to be mentioned in official mink dos.

**Pull request closed without merging**